rfcs/0003-mutations.md

```diff
@@ -0,0 +1,619 @@
+// loading: https://www.apollographql.com/docs/react/data/mutations/#tracking-mutation-status
+// state: mutation state instance
+const [exec, { loading, state }] = useMutation<MutationState<State>>(
+  CREATE_ENTITY,
+  {
+    client,
+    variables: {
+      options: { name: "...", value: 5 }
+    }
+  }
+)
+
+// Optimistic responses can be used to update
+// the UI before the execution has finished.
+// More information can be found here:
+// https://www.apollographql.com/docs/react/performance/optimistic-ui/
+const [exec, { loading, state }] = useMutation(
+  CREATE_ENTITY,
+  {
+    optimisticResponse: {
+      __typename: "Mutation",
+      createEntity: {
+        __typename: "MyEntity",
+        name: "...",
+        value: 5,
+        // NOTE: ID must be known so the
+        // final response can be correlated
+        id: "id"
+      }
+    },
+    variables: {
+      options: { name: "...", value: 5 }
+    }
+  }
+)
+```
+```html
+// Use the Mutation JSX Component
+<Mutation mutation={CREATE_ENTITY} variables={{options: { name: "...", value: 5 }}}>
+{(exec, { loading, state }) => (
+  <button onClick={exec} />
+)}
+</Mutation>
+```
+
+
+## Compatibility
+
+No breaking changes will be introduced, as mutations are an optional add-on to a subgraph.
+
+## Drawbacks and Risks
+
+Nothing apparent at the moment.
+
+## Alternatives
+
+The existing alternative that protocol developers are creating for dApp developers has been described above.
+
+## Open Questions
+
+- **How can mutations pickup where they left off in the event of an abrupt application shutdown?**
+  Since mutations can contain many different steps internally, it would be ideal to be able to support continuing resolver execution in the event the dApp abruptly shuts down.
+
+- **How can dApps understand what steps a given mutation will take during the course of its execution?**
+  dApps may way to present to the user friendly progress updates, letting them know a given action is 3/4 of the way through execution (for example) and a high level description of each step. I view this as closely tied to the previous open question above, as we could support continuing resolver executions if we know what step it's currently undergoing. More research will be done, and a potential solution to solve both of these use cases will be presented at a later date.
+
+- **Should dApps be able to define event handlers for mutation events?**
+  dApps may want to implement their own handlers for specific events emitted from mutations. These handlers would be different form the reducers, as we wouldn't want them to be able to modify the state. Instead they could store their own state elsewhere within the dApp based on the events.
+
+- **Should the resolvers module be ES5 compliant?**  
+  The prototype was originally developed under these conditions. ES5 compliance has since been abandoned as it has proven nearly impossible to successfully transpile all dependencies into a single monolithic module.
```

Comment:
It still is my hope that this is possible. But I can accept the RFC even with this question still being open.

Reply:
I was able to successfully transpile the full bundle to ES5! Removing this.

Fun Fact: If you set `mode: "production"` in your webpack config, the generated bundle is no longer ES5 compliant... not sure why this is... Was able to get ES5 compliance by setting `mode: "development"`.

Reply:
Heh. Nena may have some ideas here, but let's figure this out in the coming weeks.